building-and-packaging/material/packaging_debian_notes.md

```diff
@@ -0,0 +1,284 @@
+- We add the following changes
+    - Adding a version number to the project. This is not needed yet, but I am afraid of forgetting it later.
+    - We mark the include files needed to use the `libsse` library, i.e., `sse/sse.hpp` publicly visible.
+    - We define the target's include directory, i.e., where do we have to look for the dependency depending on the type of build (internally, top-level cmake project, external dependency)
+    - We want to make sure that we have an `install` target for make that, at the same time, considers typical installation directories. That means `make install` should binaries in a `<prefix>/bin`, libraries in `<prefix>/lib` or `<prefix>/lib/helloworld`, include files in `<prefix>/include` etc. We include the predefined CMake macro `GNUInstallDirs` to get predefined variables like the installation directories.
+
+Changes:
+
+```diff
+- project("HelloWorld")
++ project("HelloWorld" VERSION 0.1.0)
++ add_library(sse sse/sse.cpp)
++
++ set_target_properties(sse PROPERTIES PUBLIC_HEADER sse/sse.hpp)
+
++target_include_directories(sse
++    PRIVATE
++        # where the library itself will look for its internal headers
++        ${CMAKE_CURRENT_SOURCE_DIR}/helloworld
++    PUBLIC
++        # where top-level project will look for the library's public headers
++        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/helloworld>
++        # where external projects will look for the library's public headers
++        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/helloworld>
++)
+
++# Create install targets
++include(GNUInstallDirs)
++install(TARGETS helloworld sse
++  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
++  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
++  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
++  PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/helloworld
++  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/helloworld
++  )
+```
+
+- Fire up a prepared containerf or testing. Container is beneficial as it allows us to mess around with the container's paths without messing up our own system. From the example repository I run
+
+  ```bash
+  sudo docker run --rm -it --name ubuntu-packaging --mount type=bind,source="$(pwd)",target=/mnt/cpack-example jaustar/debian-package-building-base
+  ```
+
+- Build the library and install it
+
+  ```bash
+  mkdir build && cd build
+  cmake -DBUILD_SHARED_LIBS=ON ..
+  make -j
+  sudo make install
+  ```
+
+- Installs in `prefix=/usr/local`, i.e., it sets `CMAKE_INSTALL_PREFIX`.
+
+  ```bash
+  -- Install configuration: ""
+  -- Installing: /usr/local/bin/helloworld
+  -- Installing: /usr/local/lib/libsse.so
+  -- Installing: /usr/local/include/helloworld/sse.hpp
+  ```
+
+  **Note** that we have an empty configuration here. It would make sense to build the project with a proper configuration (`Debug`, `Release`...) by setting `CMAKE_BUILD_TYPE`.
```

Comment:
Why not doing it? Students know this option from last week.

Reply:
Mainly because it slipped through. I was also not sure whether I should show it first without or with the empty configuration. I will try to remember to configure everything with `-DCMAKE_BUILD_TYPE=Release` during the lecture and mention that one should always specify some build type.